src/test/fuzz/p2p_v2_transport_serialization.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2019-2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <compat/endian.h>
+#include <crypto/chacha_poly_aead.h>
+#include <crypto/poly1305.h>
+#include <key.h>
+#include <net.h>
+#include <netmessagemaker.h>
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+
+#include <cassert>
+
+FUZZ_TARGET(p2p_v2_transport_serialization)
+{
+    const CPrivKey k1(CHACHA20_POLY1305_AEAD_KEY_LEN, 0);
+    const CPrivKey k2(CHACHA20_POLY1305_AEAD_KEY_LEN, 0);
+
+    // Construct deserializer, with a dummy NodeId
+    V2TransportDeserializer deserializer{(NodeId)0, k1, k2};
+    V2TransportSerializer serializer{k1, k2};
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+
+    bool length_assist = fuzzed_data_provider.ConsumeBool();
+
+    // There is no sense is providing a mac assist if the length is incorrect.
+    bool mac_assist = length_assist && fuzzed_data_provider.ConsumeBool();
+    auto payload_bytes = fuzzed_data_provider.ConsumeRemainingBytes<uint8_t>();
+
+    if (payload_bytes.size() >= CHACHA20_POLY1305_AEAD_AAD_LEN + CHACHA20_POLY1305_AEAD_TAG_LEN) {
+        if (length_assist) {
+            uint32_t packet_length = payload_bytes.size() - CHACHA20_POLY1305_AEAD_AAD_LEN - CHACHA20_POLY1305_AEAD_TAG_LEN;
+            packet_length = htole32(packet_length);
+            memcpy(payload_bytes.data(), &packet_length, 3);
```

Comment:
I might be missing something, but, wouldn't `WriteLE32` here mean that the fourth byte (which is the first byte in the ciphertext payload) would be written over. Is there a way for me to use `WriteLE32` only for 3 bytes?

I guess that is still deterministic, so the fuzz test would work, but I'm not sure if it will work against the fuzzing engine.

Reply:
Oh you're right, here you see how easy it is to miss a off-by-one like that with memcpy! Never mind my comment. What about writing it out explicitly?
```c++
payload_bytes.data[0] = packet_length & 0xff;
payload_bytes.data[1] = (packet_length >> 8) & 0xff;
payload_bytes.data[2] = (packet_length >> 16) & 0xff;
```

Reply:
Done!